### PR TITLE
Ensure plugin load methods are called only when path/params/query change

### PR DESCRIPTION
**Pull request by bobrovnik**
Changing when plugin.routeLoad and plugin.beforeRouteLoad is happening. 

The current behavior is to load Route anytime URL changes. I'm changing it to only when Route path/params/query change.

"Resources" plugin is an exception, as it currently relies on old behavior.

### CODE REVIEW

File: src/common/utils/is-same-route/index.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import { Match } from '../../../common/types';
+
+type ObjectToCompare = {
+  [key: string]: string | null | typeof undefined;
+};
+const isDeepEqual = (object1: ObjectToCompare, object2: ObjectToCompare) => {
```
Comment (albertogasparin) [inline src/common/utils/is-same-route/index.ts:6]: Nit: this is in reality `isShallowEqual`

File: src/controllers/router-store/index.tsx [hunk 2]
```diff
@@ -149,28 +150,46 @@ const actions: AllRouterActions = {
...
 
           batch(() => {
-            plugins.forEach(p =>
-              p.beforeRouteLoad?.({
-                context: prevContext,
-                nextContext,
-              })
-            );
+            plugins.forEach(p => {
+              if (p.id === 'resources-plugin') {
```
Comment (albertogasparin) [inline src/controllers/router-store/index.tsx:165]: Can we combine the ifs?
```
if (p.id === 'resources-plugin' || !sameRoute) { ... }
```

File: src/common/utils/is-same-route/index.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { Match } from '../../../common/types';
+
+type ObjectToCompare = {
```
Comment (MonicaOlejniczak) [inline src/common/utils/is-same-route/index.ts:3]:
```suggestion
type ObjectToCompare = Record<string, string | null | undefined>;
```

File: src/common/utils/is-same-route/index.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  [key: string]: string | null | typeof undefined;
+};
+const isShallowEqual = (object1: ObjectToCompare, object2: ObjectToCompare) => {
+  const objKeys1 = Object.keys(object1);
+  const objKeys2 = Object.keys(object2);
+
+  if (objKeys1.length !== objKeys2.length) return false;
+
+  for (const key of objKeys1) {
```
Comment (MonicaOlejniczak) [inline src/common/utils/is-same-route/index.ts:12]:
```suggestion
return objKeys1.every((key) => object1[key] === object2[key]);
```

File: src/common/utils/is-same-route/index.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { Match } from '../../../common/types';
+
+type ObjectToCompare = {
+  [key: string]: string | null | typeof undefined;
+};
+const isShallowEqual = (object1: ObjectToCompare, object2: ObjectToCompare) => {
```
Comment (MonicaOlejniczak) [inline src/common/utils/is-same-route/index.ts:6]: Could just shorten it to `o1` and `o2`

**Review by at-nathan — APPROVED** [review]

**Review by albertogasparin — APPROVED** [review]

**Pull request merged**